omc/core/model/board.py

```diff
@@ -0,0 +1,545 @@
+from omc.core.exception.exception import PieceSubclassInvalidException
+
+
+class Board:
+    """
+    Object representing the state of the board, its present layout, and
+     dimensions.
+    """
+
+    def __init__(
+            self, dimensions: ndarray[int],
+            layout: ndarray[Player | None]
+    ):
+        """
+        Initialize an instance of Board.
+
+        :param dimensions: Dimensions of the board
+        :type dimensions: ndarray[int]
+        :param layout: Layout of pieces on the board
+        :type layout: ndarray[Optional[Piece]]
+        """
+        self._dimensions: ndarray[int] = dimensions
+        self._layout: ndarray[Player | None] = layout
+        self._players: dict[int, Player] = dict()
+
+    @classmethod
+    def empty(cls, dimensions: ndarray[int]) -> Board:
+        """
+        Instantiate an empty board.
+
+        :param dimensions: Dimensions of board
+        :type dimensions: ndarray[int]
+        :return: Empty board instance
+        :rtype: Board
+        """
+        return Board(dimensions, np.empty(tuple(dimensions), dtype=object))
+
+    @property
+    def dimensions(self) -> ndarray[int]:
+        """
+        Get dimensions of the board.
+
+        :return: Dimensions of the board
+        :rtype: ndarray[int]
+        """
+        return self._dimensions
+
+    @property
+    def layout(self) -> ndarray[Player | None]:
+        """
+        Get layout of the board.
+
+        :return: Layout of the board
+        :rtype: ndarray[Player | None]
+        """
+        return self._layout
+
+    @property
+    def current_players(self) -> dict[int, Player]:
+        """
+        Get current players on the board.
+
+        :return: Current players on the board
+        :rtype: ndarray[int]
+        """
+        return self._players
+
+    def add_piece(self, piece_: Piece, replace_existing: bool = False) -> None:
+        """
+        Add a piece to the board.
+
+        :param piece_: The piece to add
+        :type piece_: Piece
+        :param replace_existing: Whether to replace an existing piece at the
+            same location
+        :rtype: bool
+        :return: None
+        """
+        current_piece = cast(
+            Optional[Piece],
+            self.layout[tuple(piece_.current_coords)]
+        )
+        if current_piece and not replace_existing:
+            raise PieceAlreadyAtLocationException(
+                f'Cannot add {piece_.__class__.__name__} to the board because'
+                f' {current_piece.__class__.__name__} already present at the'
+                f' location {current_piece.current_coords}'
+            )
+        self._layout[tuple(piece_.current_coords)] = piece_
+
+    def remove_piece(self, piece_: Piece) -> None:
+        """
+        Remove a piece from the board.
+
+        :param piece_: The piece to remove
+        :type piece_: Piece
+        :return: None
+        """
+        self._layout[tuple(piece_.current_coords)] = None
```

Comment:
Not sure if its too pedantic but we may want to ensure that the piece is on the board. This would also add a boolean return type. Something like:

```
if self.query_space(piece.current_coords) != piece:
    print("Piece to remove does not belong to this board.")
    return False
```